include/aws/io/tls_channel_handler.h

```diff
@@ -706,6 +841,65 @@ AWS_IO_API struct aws_byte_buf aws_tls_handler_protocol(struct aws_channel_handl
  */
 AWS_IO_API struct aws_byte_buf aws_tls_handler_server_name(struct aws_channel_handler *handler);
 
+/**************************** TLS KEY OPERATION *******************************/
+
+/* Note: Currently this assumes the user knows what key is being used for key/cert pairs
+         but s2n supports multiple cert/key pairs. This functionality is not used in the
+         CRT currently, but in the future, we may need to implement this */
+
+/**
+ * Complete a successful TLS private key operation by providing its output.
+ * The output is copied into the TLS connection.
+ * The operation is freed by this call.
+ *
+ * You MUST call this or aws_tls_key_operation_complete_with_error().
+ * Failure to do so will stall the TLS connection indefinitely and leak memory.
+ */
+AWS_IO_API
+void aws_tls_key_operation_complete(struct aws_tls_key_operation *operation, struct aws_byte_cursor output);
+
+/**
+ * Complete an failed TLS private key operation.
+ * The TLS connection will fail.
+ * The operation is freed by this call.
+ *
+ * You MUST call this or aws_tls_key_operation_complete().
+ * Failure to do so will stall the TLS connection indefinitely and leak memory.
+ */
+AWS_IO_API
+void aws_tls_key_operation_complete_with_error(struct aws_tls_key_operation *operation, int error_code);
+
+/**
+ * Returns the input data that needs to be operated on by the custom key operation.
+ */
+AWS_IO_API
+struct aws_byte_cursor aws_tls_key_operation_get_input(const struct aws_tls_key_operation *operation);
```

Comment:
Looking at this, I'm not sure it would reduce a copy. Not that I'm against the change, I implemented it as a test but used `git stash` instead of committing it.

We are using `get_input` in two places: [s_aws_pkcs11_tls_op_handler_do_operation](https://github.com/awslabs/aws-c-io/blob/TT-key-op-prototype/source/pkcs11_tls_op_handler.c#L54) and in [s_aws_custom_key_op_handler_perform_operation](https://github.com/awslabs/aws-crt-java/blob/TT-key-op-prototype/src/native/custom_key_op_handler.c#L27).
* In the custom key op handler, we get the byte cursor and immediately call `aws_jni_byte_array_from_cursor`, which populates a JNI array with this data.
  * If we switch this to an `aws_byte_buf`, then I think we actually get an extra copy - as we have to initialize a new aws_byte_buf, populate it with the data from the operation, and then send that data to JNI using `aws_copy_native_array_to_java_byte_array` and manually pointing to pointer to it. However, under the hood, this is what `aws_jni_byte_array_from_cursor` does, **but** it does this directly from the byte cursor itself. I think if we switch this to a byte cursor, we are actually going to end up allocating an additional buffer in this case.
* In the PKCS11 handler stuff, depending on the operation we call either `aws_pkcs11_lib_decrypt` or `aws_pkcs11_lib_sign`.
  * In `aws_pkcs11_lib_decrypt` we read the data straight from the `aws_byte_cursor` itself through it's pointer and length variables. In this case, similar to the custom key operation, I think we would make an extra allocation if we use an aws_byte_buf, since it would copy the data rather than just passing the cursor around.
    * Here though, I can see a possible benefit: It would allow us to know for sure that the memory is there when we go to pass it to the PKCS11 library and we can ensure it doesn't disappear until we are done, while with the cursor, it is theoretically possible it will disappear if on other threads and such (unlikely though). It would also allow us to replace the checks using `AWS_ASSERT(cursor.len <= ULONG_MAX)` with `aws_byte_buf_is_valid`.
  * In `aws_pkcs11_lib_sign` there are different stuff happening depending if it's an RSA signature or a ECDSA. For RSA, we just get the data from the pointer and length properties in the cursor and call `s_pkcs11_sign_helper`, while in ESCA it only is used in a call to `s_pkcs11_sign_helper`. In `s_pkcs11_sign_helper`. it gets the data from the cursor using the pointer and length properties directly.
    * Like with the decrypting though, I can see the benefit of using an `aws_byte_buf` for preventing the memory from possibly being cleared when we do not expect it and for data validation, but I do not think it would remove a copy in this case either.

In both cases, I don't see how the change would reduce a copy. From what I can tell, what would just change the data type we are populating from a cursor to a buffer, which adds some additional memory management code but guarantees it will stick around.

I do not think the change is necessarily a bad thing, especially with the extra guarantees it provides for memory not disappearing if it's no another thread, but from a reducing copies perspective, based on my current experience I do not think it would do this based on what I've seen (and my admittedly limited experience).

____________

Writing this done now so I do not forget - I'm totally cool with going over this in our code review - I just don't want to forget my thoughts on it and so I wanted to capture it while it was fresh.

Reply:
good research. I'm with you